presto-hive/src/main/java/com/facebook/presto/hive/HivePartitionManager.java

```diff
@@ -143,16 +173,41 @@ public Iterable<HivePartition> getPartitionsIterator(
         else {
             return () -> {
                 List<String> partitionNames = partitionFilteringFromMetastoreEnabled ? getFilteredPartitionNames(session, metastore, hiveTableHandle, effectivePredicate) : getAllPartitionNames(session, metastore, hiveTableHandle, constraint);
-                return partitionNames.stream()
-                        // Apply extra filters which could not be done by getFilteredPartitionNames
-                        .map(partitionName -> parseValuesAndFilterPartition(tableName, partitionName, partitionColumns, partitionTypes, constraint))
-                        .filter(Optional::isPresent)
-                        .map(Optional::get)
-                        .iterator();
+
+                if (isParallelParsingOfPartitionValuesEnabled(session) && partitionNames.size() > PARTITION_NAMES_BATCH_SIZE) {
+                    List<List<String>> partitionNameBatches = Lists.partition(partitionNames, PARTITION_NAMES_BATCH_SIZE);
+                    // Use ConcurrentLinkedQueue to prevent race condition when multiple threads try to add partitions to this list
+                    ConcurrentLinkedQueue<HivePartition> result = new ConcurrentLinkedQueue<>();
+                    List<ListenableFuture<?>> futures = new ArrayList<>();
+                    try {
+                        partitionNameBatches.forEach(batch -> futures.add(executorService.submit(() -> result.addAll(getPartitionListFromPartitionNames(batch, tableName, partitionColumns, partitionTypes, constraint)))));
+                        Futures.transform(Futures.allAsList(futures), input -> result, directExecutor()).get();
+                        return result.iterator();
+                    }
+                    catch (InterruptedException | ExecutionException e) {
+                        log.error("Parallel parsing of partition values failed with error message:%s", e.getMessage());
+                    }
+                }
+                return getPartitionListFromPartitionNames(partitionNames, tableName, partitionColumns, partitionTypes, constraint).iterator();
             };
         }
     }
 
+    private List<HivePartition> getPartitionListFromPartitionNames(
+            List<String> partitionNames,
+            SchemaTableName tableName,
+            List<HiveColumnHandle> partitionColumns,
+            List<Type> partitionTypes,
+            Constraint<ColumnHandle> constraint)
+    {
+        return partitionNames.stream()
+                // Apply extra filters which could not be done by getFilteredPartitionNames
+                .map(partitionName -> parseValuesAndFilterPartition(tableName, partitionName, partitionColumns, partitionTypes, constraint))
```

Comment:
wondering if this is cacheable. Say if we have a fixed size cache targeting the > 10k partitions and if we cache `extractPartitionValues` can we get a decent cache hit rate.

Reply:
A cleaner fix is to have the Metastore send the partition values along with partition names. Then we can cache that response and get good improvements. Metastore team is working on that. tldr; yes we are working on having the values cached along with partition names.